src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs

```diff
@@ -100,11 +105,73 @@ public void Enqueue(WorkItem item)
                 {
                     Contract.ThrowIfNull(item.DocumentId);
 
-                    _highPriorityProcessor.Enqueue(item);
-                    _normalPriorityProcessor.Enqueue(item);
-                    _lowPriorityProcessor.Enqueue(item);
+                    var options = _registration.Workspace.Options;
+                    var analysisScope = SolutionCrawlerOptions.GetBackgroundAnalysisScope(options);
+
+                    if (ShouldEnqueueForAllQueues(item, analysisScope))
+                    {
+                        _highPriorityProcessor.Enqueue(item);
+                        _normalPriorityProcessor.Enqueue(item);
+                        _lowPriorityProcessor.Enqueue(item);
+                    }
+                    else
+                    {
+                        if (ShouldEnqueueReducedWorkItem(item, _highPriorityProcessor.Analyzers, options, analysisScope, out var reducedWorkItem))
+                        {
+                            _highPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
+
+                        if (ShouldEnqueueReducedWorkItem(item, _normalPriorityProcessor.Analyzers, options, analysisScope, out reducedWorkItem))
+                        {
+                            _normalPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
+
+                        if (ShouldEnqueueReducedWorkItem(item, _lowPriorityProcessor.Analyzers, options, analysisScope, out reducedWorkItem))
+                        {
+                            _lowPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
+                    }
 
                     ReportPendingWorkItemCount();
+
+                    return;
+
+                    bool ShouldEnqueueForAllQueues(WorkItem item, BackgroundAnalysisScope analysisScope)
+                    {
+                        var reasons = item.InvocationReasons;
+                        if (analysisScope == BackgroundAnalysisScope.ActiveFile &&
+                            !reasons.Contains(PredefinedInvocationReasons.DocumentClosed) &&
+                            !reasons.Contains(PredefinedInvocationReasons.DocumentRemoved))
+                        {
+                            // Only process active/closed/removed documents for active file analysis.
```

Comment:
this warrants an explanation.  your comment just describes the code, but doesn't say why.

For example, i could imagine that closing/removing the doc helps ensure that stale information goes away quickly.  but i'm just guessing, and the comment doesn't explain.

Reply:
What if a project is removed?  we don't want to similarly do this?  or is that not a problem because we'll also get individual DocumetnRemoved calls?

Reply:
Both good points - I will add more comments, test the project removed case and handle it if we are not getting individual DocumentRemoved calls.

Reply:
> What if a project is removed? we don't want to similarly do this? or is that not a problem because we'll also get individual DocumetnRemoved calls?

Verified that we do not have "ProjectRemoved" or "ProjectClosed" invocation reasons (see [here](https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/SolutionCrawler/InvocationReasons_Constants.cs)), and removing or unloading a project from solution leads to individual document removed events for all documents in the project. I have improved the comments.